pkg/ingestor/parser/parser.go

```diff
@@ -0,0 +1,286 @@
+	"github.com/guacsec/guac/pkg/handler/processor"
+	"github.com/guacsec/guac/pkg/ingestor/verifier"
+	"github.com/in-toto/in-toto-golang/in_toto"
+	"github.com/sigstore/sigstore/pkg/cryptoutils"
+)
+
+const (
+	algorithmSHA256 string = "sha256"
+)
+
+type docTreeBuilder struct {
+	identities    []assembler.IdentityNode
+	graphBuilders []*graphBuilder
+}
+
+type graphBuilder struct {
+	doc                   *processor.Document
+	foundIdentities       []assembler.IdentityNode
+	foundSubjectArtifacts []assembler.ArtifactNode
+	foundAttestations     []assembler.AttestationNode
+	foundBuilders         []assembler.BuilderNode
+	foundDependencies     []assembler.ArtifactNode
+}
+
+func newDocTreeBuilder() *docTreeBuilder {
+	return &docTreeBuilder{
+		identities:    []assembler.IdentityNode{},
+		graphBuilders: []*graphBuilder{},
+	}
+}
+
+func newGraphBuilder() *graphBuilder {
+	return &graphBuilder{
+		foundIdentities:       []assembler.IdentityNode{},
+		foundSubjectArtifacts: []assembler.ArtifactNode{},
+		foundAttestations:     []assembler.AttestationNode{},
+		foundBuilders:         []assembler.BuilderNode{},
+		foundDependencies:     []assembler.ArtifactNode{},
+	}
+}
+
+// ParseDocumentTree takes the DocumentTree and create graph inputs (nodes and edges) per document node
+func ParseDocumentTree(docTree processor.DocumentTree) ([]assembler.AssemblerInput, error) {
+
+	assemblerinputs := []assembler.AssemblerInput{}
+	docTreeBuilder := newDocTreeBuilder()
+	err := docTreeBuilder.parse(docTree)
+	if err != nil {
+		return []assembler.AssemblerInput{}, err
+	}
+	for _, builder := range docTreeBuilder.graphBuilders {
+		assemblerinput := builder.createAssemblerInput(docTreeBuilder.identities)
+		assemblerinputs = append(assemblerinputs, assemblerinput)
+	}
+
+	return assemblerinputs, nil
+}
+
+func (b *graphBuilder) createAssemblerInput(foundIdentities []assembler.IdentityNode) assembler.AssemblerInput {
+	assemblerinput := assembler.AssemblerInput{
+		Nodes: b.createNodes(),
+		Edges: b.createEdges(foundIdentities),
+	}
+	return assemblerinput
+}
+
+func (b *graphBuilder) createEdges(foundIdentities []assembler.IdentityNode) []assembler.GuacEdge {
+	edges := []assembler.GuacEdge{}
+	for _, i := range foundIdentities {
+		for _, a := range b.foundAttestations {
+			edges = append(edges, assembler.IdentityForEdge{IdentityNode: i, AttestationNode: a})
+		}
+	}
+	for _, s := range b.foundSubjectArtifacts {
+		for _, b := range b.foundBuilders {
+			edges = append(edges, assembler.BuiltByEdge{ArtifactNode: s, BuilderNode: b})
+		}
+		for _, a := range b.foundAttestations {
+			edges = append(edges, assembler.AttestationForEdge{AttestationNode: a, ArtifactNode: s})
+		}
+		for _, d := range b.foundDependencies {
+			edges = append(edges, assembler.DependsOnEdge{ArtifactNode: s, Dependency: d})
+		}
+	}
+	return edges
+}
+
+func (b *graphBuilder) createNodes() []assembler.GuacNode {
+	nodes := []assembler.GuacNode{}
+	for _, i := range b.foundIdentities {
```

Comment:
can do `nodes = append(nodes, b.foundIdentities...)`, etc.

Reply:
this throws an error hence why I did it this way.
```
cannot use b.foundIdentities (variable of type []assembler.IdentityNode) as []assembler.GuacNode value in argument to append
```
unless I am stupid and there is an easy fix to this? Tried to convert to GuacNode Type before append and it does not like that.

Reply:
can you try `append(nodes, []assembler.GuacNode(b.foundIdentities...))`? does that work?

Reply:
yea no, doesn't work nvm...